python/ql/src/experimental/semmle/python/frameworks/Xml.qll

```diff
@@ -0,0 +1,301 @@
+    }
+
+    override DataFlow::Node getAnInput() { result = this.getArg(0) }
+
+    override predicate vulnerable(string kind) {
+      exists(XML::XMLParser xmlParser |
+        xmlParser = this.getArgByName("parser").getALocalSource() and xmlParser.vulnerable(kind)
+      )
+      or
+      kind = "XXE" and not exists(this.getArgByName("parser"))
+    }
+  }
+
+  /**
+   * Gets a call to `xmltodict.parse`.
+   *
+   * Given the following example:
+   *
+   * ```py
+   * xmltodict.parse(xml_content, disable_entities=False)
+   * ```
+   *
+   * * `this` would be `xmltodict.parse(xml_content, disable_entities=False)`.
+   * * `getAnInput()`'s result would be `xml_content`.
+   * * `vulnerable(kind)`'s `kind` would be `Billion Laughs` and `Quadratic Blowup`.
+   */
+  private class XMLtoDictParsing extends DataFlow::CallCfgNode, XML::XMLParsing::Range {
+    XMLtoDictParsing() { this = API::moduleImport("xmltodict").getMember("parse").getACall() }
+
+    override DataFlow::Node getAnInput() { result = this.getArg(0) }
+
+    override predicate vulnerable(string kind) {
+      kind = ["Billion Laughs", "Quadratic Blowup"] and
+      this.getAMethodCall("disable_entities").asExpr() = any(False f)
+    }
+  }
+
+  /**
+   * Gets a call to:
+   * * `xml.dom.minidom.parse`
+   * * `xml.dom.pulldom.parse`
+   *
+   * Given the following example:
+   *
+   * ```py
+   * xml.dom.minidom.parse(StringIO(xml_content)).documentElement.childNode
+   * ```
+   *
+   * * `this` would be `xml.dom.minidom.parse(StringIO(xml_content), parser=parser)`.
+   * * `getAnInput()`'s result would be `StringIO(xml_content)`.
+   * * `vulnerable(kind)`'s `kind` would be `Billion Laughs` and `Quadratic Blowup`.
+   */
+  private class XMLDomParsing extends DataFlow::CallCfgNode, XML::XMLParsing::Range {
+    XMLDomParsing() {
+      this =
+        API::moduleImport("xml")
+            .getMember("dom")
+            .getMember(["minidom", "pulldom"])
+            .getMember(["parse", "parseString"])
+            .getACall()
+    }
+
+    override DataFlow::Node getAnInput() { result = this.getArg(0) }
+
+    override predicate vulnerable(string kind) {
+      exists(XML::XMLParser xmlParser |
+        xmlParser = this.getArgByName("parser").getALocalSource() and xmlParser.vulnerable(kind)
+      )
+      or
+      kind = ["Billion Laughs", "Quadratic Blowup"] and not exists(this.getArgByName("parser"))
+    }
+  }
+
+  /**
+   * Gets a call to `xmlrpc.server.SimpleXMLRPCServer`.
+   *
+   * Given the following example:
+   *
+   * ```py
+   * server = SimpleXMLRPCServer(("127.0.0.1", 8000))
+   * server.register_function(foo, "foo")
+   * server.serve_forever()
+   * ```
+   *
+   * * `this` would be `SimpleXMLRPCServer(("127.0.0.1", 8000))`.
+   * * `getAnInput()`'s result would be `foo`.
+   * * `vulnerable(kind)`'s `kind` would be `Billion Laughs` and `Quadratic Blowup`.
+   */
+  private class XMLRPCServer extends DataFlow::CallCfgNode, XML::XMLParser::Range {
```

Comment:
I have no idea how to fit this into the main query, as `getAnInput()` is not a `RemoteFlowSource`.